docs/spec/auth/oauth.md

```diff
@@ -188,3 +194,160 @@ Content-Type: application/json
 {"refresh_token":"kas9Da81Dfa8","access_token":"eyJhbGciOiJFUzI1NiIsInR5":"expires_in":"900","scope":"repository:samalba/my-app:pull,repository:samalba/my-app:push"}
 ````
 
+## Getting Authorization Challenge From Token Server
+
+Before authenticating with the token server the client does a `HEAD`
+request to the `/token` endpoint to get any authentication challenges.
+If the token server supports OAuth2, it will return a `WWW-Authenticate`
+with the type `OAuth2` and the parameters needed by the client to
+complete the oauth2 flow. If the client has a username and password,
+then the grant_type `password` will immediately be used. If no credentials
+were provided the daemon will pass the challenge back to the Docker client
+to complete the OAuth2 flow.
+
+### Authentication Parameters
+
+ - *client_id* - OAuth2 client id (cliend id used by provider not necessarily used by token server)
+ - *auth_url* - Authorization endpoint to send client
+ - *redirect_url* - Redirect location after login flow
+ - *landing_url* - Location to redirect after login complete
+ - *scopes* - Space separate list of scopes used by oauth provider, should be list of scopes needed registry and namespace identification.
```

Comment:
@aurelijusbanelis no `auth.docker.io` does not support this today. We would push to get that supported if we accept this proposal.

> So, I need to call auth.docker.io for every repository

On `auth.docker.io` you can actually specify the "scope" parameter multiple times to get scopes for multiple repositories. This change is trying to update it to support allow multiple in the same scope value to align with the spec, but still possible today in a non-spec compliant way.